### PR TITLE
[partially defined] implement support for try statements

### LINKED ISSUE #13928: Support try/except in partially defined check
Currently, partially defined check (`--enable-error-code partially-defined`) does not support detecting partially defined variables try/except statement. 

The most basic example is the following:
```python
try:
  x = foo()
except:
  pass
y = x  # mypy should generate an error here but does not.
```

### PR DESCRIPTION
**Pull request by ilinum**
This adds support for try/except/finally/else check.

The implementation ended up pretty complicated because it had to handle jumps different for finally. It took me a few iterations to get to this solution and that's the cleanest one I could come up with.

Let me know if you have suggestions on how to structure the code differently.

Closes #13928.

### CODE REVIEW

**Review by JukkaL — COMMENTED** [review]
Hmm, nested try statements seem to be potentially quite inefficient due to exponential behavior?  I'll try to come up with a way to avoid analyzing the same nodes multiple times.

**Comment by ilinum** [comment]
> Hmm, nested try statements seem to be potentially quite inefficient due to exponential behavior? I'll try to come up with a way to avoid analyzing the same nodes multiple times.

So right now we process the body for every except type. We could probably cache the result of processing the body and just reuse that. What do you think about that?

Note that we will still need to process it twice (once with disabling jumps and once without). That one is potentially more difficult to fix.

**Comment by JukkaL** [comment]
Here's a quick idea about how we might be able to avoid exponential behavior. I haven't fully thought this out -- there maybe some obvious thing why this doesn't work. Anyway, I wonder if we could analyze a try/except like these statements (and just once):
```py
  if ...:
      try_body
  if ...:
      except 1
  elif ...:
      except 2
  elif ...:
      except 3
  else:
      else_block
  finally_block
```

The target try statement would look like this:
```py
try:
    try_body
<except blocks 1..3>
else:
    else_block
finally:
    finally_block
```

**Comment by ilinum** [comment]
I implemented your proposal (which was pretty easy). Here's an example where this doesn't work:
```python
def f1(div: int) -> int:
    try:
        x = 100 // div
    except ArithmeticError:
        x = 0
    return x  # False positive: "x" is always defined but we get an error here.
```
This is happening because the body of the try is conditional, independently of the except clause. 

Now, I also implemented the following:
```python
if ...:
   try_body
   else_block
elif ...:  # Note the change from if to elif here.
   except 1
elif ...:
   except 2
finally_block
```

However, in this case we don't detect undefined variables inside finally itself.
```python
def f0(div: int) -> int:
    try:
        x = 100 // div
    except ArithmeticError:
        return 0
    finally:
        # "x" is not defined in the except ArithmeticError branch, which returns. 
        # Because it returns, the PartiallyDefined check considers this branch "skipped"
        # and doesn't think it's undefined.
        y = x  # False negative: this should be an error but mypy reports nothing. 
    return y
```

So to implement it 100% correct, we need to disable branch skipping. I have updated this PR with a version of the check that goes through every try/except/finally 2 times (the previous implementation was doing some unnecessary work). I believe this is the best we can do, unless we're willing to trade off correctness.

I've also found an unrelated bug in nested branch statements that get skipped. It's fixed and I added tests.

**Review by JukkaL — APPROVED** [review]
Looks good! Apologies for the slow review.

File: test-data/unit/check-partially-defined.test [hunk 1]
```diff
@@ -525,6 +525,190 @@ def f3() -> None:
...
+            x = 1
+        else:
+            return 0
+    finally:
+        a = x  # E: Name "x" may be undefined
+    return a
+[builtins fixtures/exception.pyi]
+
+[case testTryElse]
+# flags: --enable-error-code partially-defined
```
Comment (JukkaL) [inline test-data/unit/check-partially-defined.test:663]: The error codes need to be updated.

**Pull request merged**